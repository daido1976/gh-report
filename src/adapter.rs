use std::collections::HashMap;

use crate::graphql::{me::MeViewerContributionsCollection, NameWithOwner, URI};
use serde::Deserialize;

pub type MyContributions = HashMap<NameWithOwner, Vec<IssueOrPr>>;
#[derive(Debug, Deserialize)]
pub struct IssueOrPr {
    pub title: String,
    pub url: URI,
    pub state: String,
}

pub fn combine(contributions_collection: MeViewerContributionsCollection) -> MyContributions {
    let issue_contributions = contributions_collection.issue_contributions.edges.unwrap();
    let pr_contributions = contributions_collection
        .pull_request_contributions
        .edges
        .unwrap();

    let mut result: MyContributions = HashMap::new();
    for contribution in issue_contributions.into_iter().flatten() {
        let issue = contribution.node.unwrap().issue;
        let owner: NameWithOwner = issue.repository.name_with_owner;
        let issue = IssueOrPr {
            title: issue.title,
            url: issue.url,
            state: issue.state.to_string(),
        };
        result.entry(owner).or_insert_with(Vec::new).push(issue);
    }

    for contribution in pr_contributions.into_iter().flatten() {
        let pr = contribution.node.unwrap().pull_request;
        let owner: NameWithOwner = pr.repository.name_with_owner;
        let pr = IssueOrPr {
            title: pr.title,
            url: pr.url,
            state: pr.state.to_string(),
        };
        result.entry(owner).or_insert_with(Vec::new).push(pr);
    }
    result
}

#[cfg(test)]
mod test {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_combine() {
        let cc =
            serde_json::from_str(include_str!("fixtures/contributionsCollection.json")).unwrap();
        let actual = combine(cc);
        assert_debug_snapshot!(actual);
    }
}
