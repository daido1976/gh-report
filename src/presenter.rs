use crate::adapter;
use std::fmt::Write;

pub fn to_string_pretty(my_contributions: adapter::MyContributions) -> String {
    let mut result = String::new();
    for (owner, issue_or_prs) in my_contributions {
        let _ = writeln!(result, "\n### {}\n", owner);
        for issue_or_pr in issue_or_prs {
            let _ = writeln!(
                result,
                "- [{}]({}) **{}!**",
                issue_or_pr.title, issue_or_pr.url, issue_or_pr.state
            );
        }
    }
    result
}

#[cfg(test)]
mod test {
    use super::*;
    use insta::assert_display_snapshot;

    #[test]
    fn test_to_string_pretty() {
        let cc = serde_json::from_str(include_str!("fixtures/myContributions.json")).unwrap();
        let actual = to_string_pretty(cc);
        assert_display_snapshot!(actual);
    }
}
