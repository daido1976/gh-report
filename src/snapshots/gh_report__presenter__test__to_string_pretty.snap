---
source: src/presenter.rs
assertion_line: 28
expression: actual
---

### rust-lang/rust

- [Cannot build on Fedora: wrong CPUTYPE?](https://github.com/rust-lang/rust/issues/1218) **open!**
- [rustc: Fix position of diagnostic highlight lines](https://github.com/rust-lang/rust/issues/1219) **closed!**
- [Add float support to #fmt.](https://github.com/rust-lang/rust/pull/1168) **merged!**
- [Prohibit in-scope consts from use as variable names in binders, like nullary tags](https://github.com/rust-lang/rust/pull/1193) **merged!**

### daido1976/terakoya

- [From now to Vercel](https://github.com/daido1976/terakoya/pull/52) **merged!**
